kuka_rsi_hw_interface/src/rsi_command_handler.cpp

```diff
@@ -0,0 +1,371 @@
+  state_data_structure_.Childs().emplace_back(ASPos_el);
+  state_data_structure_.Childs().emplace_back(Delay_el);
+  state_data_structure_.Childs().emplace_back(Ipoc_state_el);
+
+  // Command structure
+  command_data_structure_.Params()["Type"] = xml::XMLParam(xml::XMLType::STRING);
+  command_data_structure_.SetParam<xml::XMLString>("Type", xml::XMLString("KROSHU"));
+  xml::XMLElement ak_el("AK");
+  ak_el.Params()["A1"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  ak_el.Params()["A2"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  ak_el.Params()["A3"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  ak_el.Params()["A4"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  ak_el.Params()["A5"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  ak_el.Params()["A6"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  xml::XMLElement Stop_el("Stop");
+  Stop_el.Params()["Stop"] = xml::XMLParam(xml::XMLType::BOOL);
+  xml::XMLElement Ipoc_command_el("IPOC");
+  Ipoc_command_el.Params()["IPOC"] = xml::XMLParam(xml::XMLType::LONG);
+  command_data_structure_.Childs().emplace_back(ak_el);
+  command_data_structure_.Childs().emplace_back(Stop_el);
+  command_data_structure_.Childs().emplace_back(Ipoc_command_el);
+}
+
+bool RSICommandHandler::Decode(char * const buffer, const size_t buffer_size)
+{
+  auto buffer_it = buffer;
+  try {
+    decodeNodes(this->state_data_structure_, buffer, buffer_it, buffer_size);
+    return true;
+  } catch (const std::exception & e) {
+    RCLCPP_ERROR(rclcpp::get_logger("CommandHandler"), "%s", e.what());
+    return false;
+  }
+}
+
+int RSICommandHandler::Encode(char * & buffer, const size_t buffer_size)
+{
+  auto buffer_it = buffer;
+  int size_left = buffer_size;
+  try {
+    encodeNodes(this->command_data_structure_, buffer_it, size_left);
+    if (static_cast<int>(buffer_size) - size_left != static_cast<int>(buffer_it - buffer)) {
+      throw std::range_error("Range error occured");
+    }
+    // +1 is for the \0 character
+    return buffer_size - size_left + 1;
+  } catch (const std::exception & e) {
+    RCLCPP_ERROR(rclcpp::get_logger("CommandHandler"), "%s", e.what());
+    return -1;
+  }
+}
+
+void RSICommandHandler::decodeNodes(
+  xml::XMLElement & element, char * const buffer, char * & buffer_it,
+  const size_t buffer_size)
+{
+  xml::XMLString node_name(buffer, 0);
+  if (*(buffer_it) != '<') {
+    std::stringstream err_ss;
+    err_ss << "Syntax error at the start of " << element.GetName() << " node.";
+    throw std::logic_error(err_ss.str());
+  }
+  buffer_it++;
+  // Validate nodes name
+  if (!element.IsNameValid(node_name, buffer_it)) {
+    std::stringstream err_ss;
+    err_ss << "The detected start node name \"" << node_name <<
+      "\" does not match the elements name " <<
+      element.GetName() << ".";
+    throw std::logic_error(err_ss.str());
+  }
+
+  // Validate nodes params
+  size_t num_of_param = 0;
+  bool is_param_only_node = false;
+  bool isNoMoreParam = false;
```

Comment:
could you also change this to snake_case?

Reply:
done